File: 06_strings/06_11_longest_palindromic_subsequence/longest_palindromic_subsequence.cpp
#include <iostream>

std::string extendPalindrome(const std::string& s, int left, int right) {
    while (left < s.length() && right >= 0 && s[left] == s[right]) {
        left++;
        right--;
    }
    return s.substr(right + 1, left - right - 1);
}

std::string longestPalindrome(const std::string& s) {
    std::string result = "";
    for (int i = 0; i < s.length(); i++) {
        std::string temp = extendPalindrome(s, i, i);
        if (temp.length() > result.length()) {
            result = temp;
        }
        temp = extendPalindrome(s, i + 1, i);
        if (temp.length() > result.length()) {
            result = temp;
        }
    }
    return result;
}

int main() {
    // Test cases
    std::cout << "Input: \"babad\" -> Output: \"" << longestPalindrome("babad") << "\"" << std::endl;
    std::cout << "Input: \"cbbd\"  -> Output: \"" << longestPalindrome("cbbd") << "\"" << std::endl;
    std::cout << "Input: \"a\"     -> Output: \"" << longestPalindrome("a") << "\"" << std::endl;
    std::cout << "Input: \"ac\"    -> Output: \"" << longestPalindrome("ac") << "\"" << std::endl;

    return 0;
}

File: 06_strings/06_11_longest_palindromic_subsequence/LongestPalindromicSubsequence.java
public class LongestPalindromicSubsequence {
    public static void main(String[] args) {
        // Test cases
        test("babad"); // Output: "bab" or "aba"
        test("cbbd");  // Output: "bb"
        test("a");     // Output: "a"
        test("ac");    // Output: "a" or "c"
    }

    private static void test(String input) {
        String result = longestPalindrome(input);
        System.out.println("Input: \"" + input + "\" -> Output: \"" + result + "\"");
    }

    private static String longestPalindrome(String s) {
        String result = "";
        for (int i = 0; i < s.length(); i++) {
            String temp = extendPalindrome(s, i, i);
            if (temp.length() > result.length()) {
                result = temp;
            }
            temp = extendPalindrome(s, i + 1, i);
            if (temp.length() > result.length()) {
                result = temp;
            }
        }
        return result;
    }

    private static String extendPalindrome(String s, int left, int right) {
        while (left < s.length() && right >= 0 && s.charAt(left) == s.charAt(right)) {
            left++;
            right--;
        }
        return s.substring(right + 1, left);
    }
}
